netutil/netutil.go

```diff
@@ -0,0 +1,129 @@
+package netutil
+
+import (
+	"fmt"
+	"net"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+)
+
+func ipForAddr(addr net.Addr) (net.IP, bool) {
+	switch a := addr.(type) {
+	case *net.IPAddr:
+		return a.IP, true
+	case *net.IPNet:
+		return a.IP, true
+	default:
+		return net.IPv4zero, false
+	}
+}
+
+func PrivateNetworkInterfaces() []string {
+	ifaces := []string{}
+
+	all, err := net.Interfaces()
+	if err != nil {
+		return ifaces
+	}
+
+IFACES:
+	for _, iface := range all {
+		if iface.Flags&net.FlagLoopback == 0 && iface.Flags&net.FlagUp != 0 {
+			addrs, err := iface.Addrs()
+			if err != nil {
+				continue
+			}
+			for _, addr := range addrs {
+				ip, ok := ipForAddr(addr)
+				if !ok {
+					continue
+				}
+				if !IsPrivate(ip) {
+					continue IFACES
+				}
+			}
+			ifaces = append(ifaces, iface.Name)
+		}
+	}
+	return ifaces
+}
+
+// FirstAddressOf returns the first IPv4 address of the supplied interface
+// names, omitting any 169.254.x.x automatic private IPs if possible.
+func FirstAddressOf(names []string, logger log.Logger) (string, error) {
+	var ipAddr net.IP
+	for _, name := range names {
+		inf, err := net.InterfaceByName(name)
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting interface", "inf", name, "err", err)
+			continue
+		}
+		addrs, err := inf.Addrs()
+		if err != nil {
+			level.Warn(logger).Log("msg", "error getting addresses for interface", "inf", name, "err", err)
+			continue
+		}
+		if len(addrs) <= 0 {
+			level.Warn(logger).Log("msg", "no addresses found for interface", "inf", name, "err", err)
+			continue
+		}
+		if ip := filterIPs(addrs); !ip.IsUnspecified() {
+			ipAddr = ip
+		}
+		if isAPIPA(ipAddr) || ipAddr.IsUnspecified() {
+			continue
+		}
+		return ipAddr.String(), nil
+	}
+	if ipAddr.IsUnspecified() {
+		return "", fmt.Errorf("no address found for %s", names)
+	}
+	if isAPIPA(ipAddr) {
+		level.Warn(logger).Log("msg", "using automatic private ip", "address", ipAddr)
```

Comment:
I'd also mention why we fall back to using an automatic private ip
```suggestion
		level.Warn(logger).Log("msg", "using automatic private ip because no other one was found", "address", ipAddr)
```

Reply:
If there are multiple automatic private addresses and no other addresses then this will return the last one of them, which I think is confusing given the name of the function being `FirstAddressOf`